packages/data/src/test/privateAPIs.js

```diff
@@ -164,6 +165,43 @@ describe( 'Private data APIs', () => {
 			);
 			expect( subPrivateSelectors.getSecretDiscount() ).toEqual( 800 );
 		} );
+
+		it( 'should support registry selectors', () => {
+			const groceryStore = createStore();
+			const otherStore = createReduxStore( 'other', {
+				reducer: ( state = {} ) => state,
+			} );
+			registry.register( otherStore );
+			unlock( otherStore ).registerPrivateSelectors( {
+				getPrice: createRegistrySelector(
+					( select ) => () => select( groceryStore ).getPublicPrice()
+				),
+			} );
+			const privateSelectors = unlock( registry.select( otherStore ) );
+			expect( privateSelectors.getPrice() ).toEqual( 1000 );
+		} );
+
+		it( 'should support calling a private registry selector from a public selector', () => {
+			const groceryStore = createStore();
+			const storeA = createReduxStore( 'a', {
+				reducer: ( state = {} ) => state,
+			} );
+			registry.register( storeA );
+			const getPrice = createRegistrySelector(
+				( select ) => () => select( groceryStore ).getPublicPrice()
+			);
+			unlock( storeA ).registerPrivateSelectors( {
+				getPrice,
+			} );
+			const storeB = createReduxStore( 'b', {
+				reducer: ( state = {} ) => state,
+				selectors: {
+					getPrice: ( state ) => getPrice( state ),
```

Comment:
This test it bad, you can't really expect that it will work.
- a registry selector must be registered with a store to "activate" it, i.e., to set the `selector.registry` reference. Because in `storeA` it's only registered lazily, and never called, it's not activated yet.
- in `storeB` you should register it directly, with `selectors: { getPrice }`. Then the test would work.
- because the implemenation sets a `.registry` property on the `getPrice` object  returned from `createRegistrySelector`, you should really use `getPrice` just once. And not register it in two stores. Otherwise both usages will be fighting for the `selector.registry` field and the results can in theory be unpredictable.

Where is the real-world situation where you're running into this kind of issue?

Reply:
Thanks for clarifying the test. In the previous version, behavior of store B depended on what was registered in store A, and that was fishy 🙂 

I cherry-picked the tests into #51051 and fixed it there by pre-binding private selectors on store instantiation.

I had to modify the test so that the private selector registration is done _before_ the store is instantiated by `registry.register`. Otherwise we don't have a good opportunity to pre-bind selectors.

The `registerPrivateSelectors` API is weird in the sense that it allows registering selectors i) multiple times, and ii) after the store has been instantianted one or more times, and iii) the late registration affects also all the instances that already exist, it is retroactive!

That should probably also be eventually fixed. We'll get to that sooner or later.

@adamziel suggests this:
```js
createReduxStore( 'a', {
  selectors: lock( { foo }, { privateFoo } )
} );
```
It's syntactically elegant, but semantically it's nonsense 🙂 To register a private selector, we need to `unlock` a private API of `@wordpress/data`, but here we're achieving that by _locking_.

What we'd like to do is:
1) Register private selectors inside the `createReduxStore` call, so that the entire registration is atomic, and `createReduxStore` returns a store descriptor that is immutable.
2) Expose private selectors registration only as a private API, and that means having to call `unlock` somewhere.

One way to satisfy that could be:
```js
unlock( createReduxStore )( 'a', {
  selectors: { foo },
  privateSelectors: { privateFoo },
} );
```
Are there others?

Reply:
A private `privateSelectors` option makes complete sense to me.